File: PhotoPicker/src/main/java/hai/ithust/photopicker/fragment/PhotoPickerFragment.java
package hai.ithust.photopicker.fragment;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.OrientationHelper;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import hai.ithust.photopicker.PhotoPicker;
import hai.ithust.photopicker.R;
import hai.ithust.photopicker.adapter.PhotoGridAdapter;
import hai.ithust.photopicker.entity.GalleryPhoto;
import hai.ithust.photopicker.entity.PhotoDirectory;
import hai.ithust.photopicker.event.OnPhotoListener;
import hai.ithust.photopicker.utils.ImageCaptureManager;
import hai.ithust.photopicker.utils.PermissionsConstant;
import hai.ithust.photopicker.utils.PermissionsUtils;

import static android.app.Activity.RESULT_OK;
import static hai.ithust.photopicker.PhotoPicker.DEFAULT_COLUMN_NUMBER;
import static hai.ithust.photopicker.PhotoPicker.DEFAULT_MAX_COUNT;
import static hai.ithust.photopicker.PhotoPicker.EXTRA_MAX_COUNT;

/**
 * @author conghai on 12/20/18.
 */
public class PhotoPickerFragment extends Fragment implements OnPhotoListener, PhotoPickerCallback {
    private PhotoPickerPresenter mPresenter;

    private ImageCaptureManager mCaptureManager;
    private PhotoGridAdapter mPhotoAdapter;

    private AlertDialog mSelectDirectoryDialog;
    private TextView mTvTile;
    private ImageView mIvRightAction;

    public static PhotoPickerFragment newInstance(Bundle bundle) {
        PhotoPickerFragment fragment = new PhotoPickerFragment();
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            ArrayList<String> originalPhotos = getArguments().getStringArrayList(PhotoPicker.EXTRA_ORIGINAL_PHOTOS);
            int maxCount = getArguments().getInt(EXTRA_MAX_COUNT, DEFAULT_MAX_COUNT);

            mPhotoAdapter = new PhotoGridAdapter(originalPhotos, maxCount, this);
            mCaptureManager = new ImageCaptureManager(getActivity());
        }
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mPresenter = new PhotoPickerPresenter(this);
        mPresenter.create();
        mPresenter.getPhotos(getContext());
    }

    @Override
    public void onDestroyView() {
        if (mSelectDirectoryDialog != null) {
            mSelectDirectoryDialog.dismiss();
        }
        mPresenter.destroy();
        super.onDestroyView();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.picker_fragment_photo_picker, container, false);
        int column = getArguments() != null ? getArguments().getInt(PhotoPicker.EXTRA_GRID_COLUMN, DEFAULT_COLUMN_NUMBER) : DEFAULT_COLUMN_NUMBER;

        RecyclerView recyclerView = rootView.findViewById(R.id.rv_photos);
        StaggeredGridLayoutManager layoutManager = new StaggeredGridLayoutManager(column, OrientationHelper.VERTICAL);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setItemAnimator(null);
        recyclerView.setHasFixedSize(true);
        recyclerView.setAdapter(mPhotoAdapter);

        mTvTile = rootView.findViewById(R.id.tv_directory_category);
        mIvRightAction = rootView.findViewById(R.id.iv_right_action);

        mTvTile.setOnClickListener(v -> {
            if (isResumed()) {
                mSelectDirectoryDialog = getDirectoryDialog(mPhotoAdapter.getDirectoryCategoryIndex());
                mSelectDirectoryDialog.show();
            }
        });

        mIvRightAction.setOnClickListener(view -> {
            if (mPhotoAdapter != null && mPhotoAdapter.isSelectPhoto()) {
                setResult();
            } else {
                openCamera();
            }
        });

        rootView.findViewById(R.id.iv_navigation).setOnClickListener(view -> {
            if (getActivity() != null) {
                getActivity().finish();
            }
        });

        return rootView;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == ImageCaptureManager.REQUEST_TAKE_PHOTO && resultCode == RESULT_OK) {
            if (mCaptureManager == null) {
                mCaptureManager = new ImageCaptureManager(getContext());
            }

            mCaptureManager.galleryAddPic();
            String path = mCaptureManager.getCurrentPhotoPath();
            if (mPhotoAdapter != null && !TextUtils.isEmpty(path)) {
                mPhotoAdapter.addNewPhoto(new GalleryPhoto(path.hashCode(), path));
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (grantResults.length > 0
                && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            switch (requestCode) {
                case PermissionsConstant.REQUEST_CAMERA:
                case PermissionsConstant.REQUEST_EXTERNAL_WRITE:
                    if (PermissionsUtils.checkWriteStoragePermission(this) &&
                            PermissionsUtils.checkCameraPermission(this)) {
                        openCamera();
                    }
                    break;
            }
        }
    }


    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        mCaptureManager.onSaveInstanceState(outState);
        super.onSaveInstanceState(outState);
    }


    @Override
    public void onViewStateRestored(Bundle savedInstanceState) {
        mCaptureManager.onRestoreInstanceState(savedInstanceState);
        super.onViewStateRestored(savedInstanceState);
    }

    @Override
    public void onItemCheck(View view, int position) {
        if (mPhotoAdapter != null) {
            boolean isUpdated = mPhotoAdapter.updatePhoto(position);

            if (mPhotoAdapter.isSelectPhoto()) {
                mIvRightAction.setImageResource(R.drawable.picker_ic_done);
            } else {
                mIvRightAction.setImageResource(R.drawable.picker_ic_camera);
            }
            if (!isUpdated) {
                Toast.makeText(getContext(), getString(R.string.__picker_over_max_count_tips, mPhotoAdapter.getMaxItem()), Toast.LENGTH_SHORT).show();
            }
        }
    }

    @Override
    public void onOpenCamera() {
        if (PermissionsUtils.checkWriteStoragePermission(this) &&
                PermissionsUtils.checkCameraPermission(this)) {
            openCamera();
        }
    }

    @Override
    public void onGetListPhotoSuccess(List<PhotoDirectory> directories) {
        mPhotoAdapter.setPhotoDirectories(directories);
    }

    private void openCamera() {
        try {
            Intent intent = mCaptureManager.dispatchTakePictureIntent();
            startActivityForResult(intent, ImageCaptureManager.REQUEST_TAKE_PHOTO);
        } catch (Exception ignore) {
        }
    }

    private void setResult() {
        if (getActivity() != null) {
            Intent intent = new Intent();
            intent.putStringArrayListExtra(PhotoPicker.KEY_SELECTED_PHOTOS, mPhotoAdapter.getSelectedPhotos());
            getActivity().setResult(RESULT_OK, intent);
            getActivity().finish();
        }
    }

    private AlertDialog getDirectoryDialog(int position) {
        List<String> directories = mPhotoAdapter.getPhotoDirectories();
        directories.add(0, getString(R.string.pp_gallery));
        CharSequence[] items = directories.toArray(new CharSequence[0]);
        AlertDialog.Builder builder = new AlertDialog.Builder(Objects.requireNonNull(getActivity()))
                .setTitle(R.string.pp_select_directory_title)
                .setSingleChoiceItems(items, position, (dialog, which) -> {
                    mTvTile.setText(items[which]);
                    mPhotoAdapter.setCurrentDirectoryIndex(which);

                    dialog.dismiss();
                });
        return builder.create();
    }
}

File: PhotoPicker/src/main/java/hai/ithust/photopicker/fragment/PhotoPickerFragment.kt
package hai.ithust.photopicker.fragment

import android.app.Activity
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.OrientationHelper
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import hai.ithust.photopicker.PhotoPicker
import hai.ithust.photopicker.R
import hai.ithust.photopicker.adapter.PhotoGridAdapter
import hai.ithust.photopicker.entity.GalleryPhoto
import hai.ithust.photopicker.entity.PhotoDirectory
import hai.ithust.photopicker.event.OnPhotoListener
import hai.ithust.photopicker.utils.ImageCaptureManager
import hai.ithust.photopicker.utils.PermissionsConstant
import hai.ithust.photopicker.utils.PermissionsUtils.checkCameraPermission
import hai.ithust.photopicker.utils.PermissionsUtils.checkWriteStoragePermission

/**
 * @author conghai on 12/20/18.
 */
class PhotoPickerFragment : Fragment(R.layout.picker_fragment_photo_picker), OnPhotoListener, PhotoPickerCallback {
    private lateinit var tvTitle: TextView
    private lateinit var ivRightAction: ImageView
    private lateinit var presenter: PhotoPickerPresenter
    private lateinit var photoAdapter: PhotoGridAdapter
    private lateinit var captureManager: ImageCaptureManager

    private var mSelectDirectoryDialog: AlertDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        presenter = PhotoPickerPresenter(this)
        val originalPhotos = requireArguments().getParcelableArrayList<Uri>(PhotoPicker.EXTRA_ORIGINAL_PHOTOS)
        val maxCount = requireArguments().getInt(PhotoPicker.EXTRA_MAX_COUNT, PhotoPicker.DEFAULT_MAX_COUNT)
        photoAdapter = PhotoGridAdapter(originalPhotos, maxCount, this)
        captureManager = ImageCaptureManager(requireContext())
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val column = requireArguments().getInt(PhotoPicker.EXTRA_GRID_COLUMN, PhotoPicker.DEFAULT_COLUMN_NUMBER)
        val recyclerView: RecyclerView = view.findViewById(R.id.rv_photos)

        recyclerView.apply {
            layoutManager = StaggeredGridLayoutManager(column, OrientationHelper.VERTICAL)
            itemAnimator = null
            setHasFixedSize(true)
            adapter = photoAdapter
        }

        tvTitle = view.findViewById(R.id.tv_directory_category)
        ivRightAction = view.findViewById(R.id.iv_right_action)

        tvTitle.setOnClickListener {
            if (isResumed) {
                mSelectDirectoryDialog = getDirectoryDialog(photoAdapter.directoryCategoryIndex)
                mSelectDirectoryDialog!!.show()
            }
        }

        ivRightAction.setOnClickListener {
            if (photoAdapter.isSelectPhoto) {
                setResult()
            } else {
                onOpenCamera()
            }
        }

        view.findViewById<View>(R.id.iv_navigation).setOnClickListener {
            activity?.finish()
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        presenter.getPhotos(requireContext())
    }

    override fun onDestroyView() {
        mSelectDirectoryDialog?.dismiss()
        presenter.destroy()
        super.onDestroyView()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == ImageCaptureManager.REQUEST_TAKE_PHOTO && resultCode == Activity.RESULT_OK) {
            captureManager.galleryAddPic()
            captureManager.currentUri?.let {
                photoAdapter.addNewPhoto(GalleryPhoto(it))
            }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            when (requestCode) {
                PermissionsConstant.REQUEST_CAMERA, PermissionsConstant.REQUEST_EXTERNAL_WRITE -> onOpenCamera()
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        captureManager.onSaveInstanceState(outState)
        super.onSaveInstanceState(outState)
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        captureManager.onRestoreInstanceState(savedInstanceState)
        super.onViewStateRestored(savedInstanceState)
    }

    override fun onItemCheck(position: Int) {
        val isUpdated = photoAdapter.updatePhoto(position)
        updateRightAction()
        if (!isUpdated) {
            Toast.makeText(context, getString(R.string.__picker_over_max_count_tips, photoAdapter.maxItem), Toast.LENGTH_SHORT).show()
        }
    }

    override fun onOpenCamera() {
        if (checkWriteStoragePermission(this) &&
                checkCameraPermission(this)) {
            openCamera()
        }
    }

    override fun onGetListPhotoSuccess(directories: List<PhotoDirectory>) {
        photoAdapter.setPhotoDirectories(directories)
        updateRightAction()
    }

    private fun openCamera() {
        try {
            val intent = captureManager.dispatchTakePictureIntent()
            startActivityForResult(intent, ImageCaptureManager.REQUEST_TAKE_PHOTO)
        } catch (ignore: Exception) {
        }
    }

    private fun setResult() {
        val intent = Intent()
        intent.putParcelableArrayListExtra(PhotoPicker.KEY_SELECTED_PHOTOS, photoAdapter.selectedPhotos)
        if (targetFragment != null) {
            targetFragment!!.onActivityResult(targetRequestCode, Activity.RESULT_OK, intent)
        } else {
            requireActivity().setResult(Activity.RESULT_OK, intent)
            requireActivity().finish()
        }
    }

    private fun getDirectoryDialog(position: Int): AlertDialog {
        val directories = mutableListOf(getString(R.string.pp_gallery))
        directories.addAll(photoAdapter.photoDirectoryNames)
        val items = directories.toTypedArray<CharSequence>()
        val builder = AlertDialog.Builder(requireContext())
                .setTitle(R.string.pp_select_directory_title)
                .setSingleChoiceItems(items, position) { dialog: DialogInterface, which: Int ->
                    tvTitle.text = items[which]
                    photoAdapter.setCurrentDirectoryIndex(which)
                    dialog.dismiss()
                }
        return builder.create()
    }

    private fun updateRightAction() {
        if (photoAdapter.isSelectPhoto) {
            ivRightAction.setImageResource(R.drawable.picker_ic_done)
        } else {
            ivRightAction.setImageResource(R.drawable.picker_ic_camera)
        }
    }

    companion object {
        @JvmStatic
        fun newInstance(bundle: Bundle?): PhotoPickerFragment {
            return PhotoPickerFragment().apply {
                arguments = bundle
            }
        }
    }
}